### ISSUE TITLE
[P1][generation] explicit runner contract and removal of signature guessing / runtime class swapping

### ISSUE THREAD

**Issue by nishide-dev** [nishide-dev/unturtle#186]
## Depends on

- #184 architecture contract characterization

May proceed in parallel with #185 after #184 freezes the current generation execution map. Do not mix loader/PEFT extraction into this chain.

## Goal

Replace implicit generation invocation with an explicit runner contract while preserving each family's native state, config and sampling semantics.

The desired boundary is:

- registry selects an algorithm;
- an explicit adapter invokes that algorithm for the model family;
- family-specific `GenerationConfig` and outputs remain authoritative;
- callers may use a canonical Unturtle dispatch function without requiring Unturtle to mutate the loaded model's class;
- existing model `.generate(...)` entry points remain compatible during migration.

## Why

The current selection registry is directionally correct: it separates masked diffusion, canvas diffusion, continuous flow, flow map and latent paths without inventing a universal solver/config.

The weak parts are execution seams:

1. `_call_sampling_loop` inspects private method signatures, guesses parameter meaning from names, and passes `None` for missing required values;
2. support is often inferred from the presence of a private method such as `_sample` or `_denoising_step`;
3. DiffusionGemma gains Unturtle's `algorithm=` shim through runtime `__class__` replacement;
4. an instance-level Unsloth `generate` override is deleted so the swapped class method becomes visible;
5. runner output and executed NFE/steps are not represented by one explicit result contract.

These mechanisms are difficult to reason about under upstream Transformers/Unsloth changes and can make the selected algorithm differ from the method that actually executed.

## Frozen design constraints

1. **No universal generation config.** Dream/LLaDA/Tiny-A2D/MDLM-DiT/DiffusionGemma/ELF/FMLM keep their native configs and semantics.
2. **No universal solver/state abstraction.** The runner boundary selects/invokes; it does not reinterpret the method.
3. **No paper-name branches in central dispatch.** Family adapters register through the same hub contract as builtins/plugins.
4. **No capability promotion.** A registered runner is not proof that every model can execute it.
5. **Preserve `algorithm="auto"` ordering and explicit unsupported errors** unless a separately reviewed contract change is required.
6. **Preserve #157 benchmark semantics.** Infrastructure parity is not a systems speedup and must not be credited as one.

## Target contract

Use a small protocol or frozen dataclass-backed callable. Suggested shape:

```python
class GenerationRunner(Protocol):
    name: str
    family: str

    def support(self, model: object) -> SupportResult: ...
    def run(
        self,
        model: object,
        request: GenerationRequest,
    ) -> GenerationResult: ...
```

### `SupportResult`

Must distinguish:

- supported;
- unsupported with typed reason and human message;
- unverified because a required optional/provider integration is unavailable.

It may include structural requirements such as:

- required callable/adapter;
- topology incompatibility;
- prompt support;
- cache support;
- native provider identity.

Do not make public capability strings the sole runtime predicate.

### `GenerationRequest`

Keep the current family-opaque intent:

```python
GenerationRequest(
    inputs=...,
    generation_config=...,
    kwargs={...},
)
```

Add only fields proven to have at least two real consumers. Do not add masked-diffusion flags to the common request.

### `GenerationResult`

Minimum common envelope:

```python
GenerationResult(
    output=...,
    algorithm=...,
    family=...,
    provider=...,
    execution={...},
)
```

`execution` may carry stable common measurements when available:

- executed model evaluations/NFE;
- executed steps;
- selected runner identity;
- native result type;
- typed unsupported/OOM when the caller requests record production.

Method-specific traces remain tagged/provider-specific. Do not force every method to emit commit order, solver state or latent diagnostics.

## Canonical dispatch

Introduce or make explicit a model-external call path, for example:

```python
unturtle.generate(
    model,
    inputs,
    algorithm="auto",
    generation_config=config,
    hub=hub,
    **kwargs,
)
```

Naming/location should follow #184's public-surface verdict. The load-bearing requirement is **external dispatch without class mutation**, not the exact function name.

Existing `.generate(algorithm=...)` mixins/shims should delegate to the same runner path or remain as compatibility adapters. There must be one selection implementation, not two drifting algorithms.

## Builtin adapters

Create explicit adapters for current builtin execution families.

### Masked MDLM

- calls the exact masked sampling loop with explicitly mapped arguments;
- no signature inspection;
- records executed steps/NFE from the loop or instrumentation-owned counter;
- supports the same prompt/attention-mask semantics as before.

### Block decode

- calls the exact cache sampling path;
- preserves hybrid-topology rejection;
- explicit cache/prompt requirements;
- no fallback to MDLM after explicit selection.

### BD3LM

- explicit Tiny-A2D/block-diffusion adapter;
- remains opt-in and not automatically promoted.

### DFM / FlowLM / LaDiff prototypes

- retain research-only opt-in predicates;
- reject unsupported prompts loudly;
- pack/model-specific kwargs remain opaque to core.

### DiffusionGemma canvas generation

- use the loaded upstream Transformers model directly;
- select `block_ar` through an external/native adapter;
- delegate to upstream `generate` without runtime `__class__` replacement;
- do not delete or overwrite an existing instance-level `generate` attribute;
- preserve exact upstream `DiffusionGemmaGenerationConfig` behavior;
- prove output and execution parity against the pre-refactor wrapper path before removing the swap.

If Unsloth's instance wrapper changes incompatible kwargs/cache behavior, resolve that at the adapter/provider boundary with an explicit callable reference or upstream delegation—not by changing the object's class.

## Plugin/method-pack contract

Plugins should be able to register a runner/algorithm against an explicit hub without importing or mutating the default hub.

A plugin adapter must provide:

- selection metadata;
- support result;
- explicit runner callable;
- provider provenance through the existing plugin mechanism.

Do not require external method packs to inherit a core generation mixin if a registered adapter is sufficient.

### PR 0 — characterize and introduce result/support types

- consume #184 execution map;
- add `SupportResult` / `GenerationResult` or equivalent;
- retain old runner internals;
- prove selected algorithm and actual method are recorded independently.

### PR 1 — explicit masked-family adapters

- MDLM, block decode, BD3LM;
- remove signature inspection for migrated adapters;
- differential fixed-seed parity;
- exact unsupported-message parity where public.

### PR 2 — research/continuous adapters

- DFM, FlowLM, LaDiff and any builtin research-only paths;
- maintain opt-in/capability boundaries;
- no central state/solver abstraction.

### PR 3 — external dispatch surface

- one canonical dispatch function using an explicit hub;
- existing model mixins delegate or adapt to it;
- no import-time plugin discovery;
- isolated hubs remain independent.

### PR 4 — DiffusionGemma class-swap removal

- add direct native adapter;
- parity-test upstream output/config/RNG behavior;
- remove `_POST_LOAD_CLASS_SWAPS` dependency for DiffusionGemma;
- remove mutation/deletion of instance `generate` only after parity;
- ensure old `model.generate(..., algorithm="block_ar")` compatibility is addressed explicitly: retained via safe wrapper/adapter or deprecated with a separately documented path. Do not silently drop it.

### PR 5 — delete legacy signature guessing

Only after all builtin algorithms and tested plugins use explicit adapters:

- remove `_call_sampling_loop` or leave a clearly deprecated plugin-compatibility shim with telemetry/tests;
- no new registrations may use the legacy path.

## Differential and semantic requirements

For each migrated algorithm:

- requested algorithm equals recorded selected runner;
- exact private/native method invoked is known;
- same deterministic output under fixed inputs/config/RNG where semantics permit;
- same RNG ownership and terminal state;
- same executed steps/NFE;
- same prompt/attention-mask behavior;
- same explicit unsupported behavior;
- same Auto priority and opt-in rules;
- no hidden fallback after explicit selection;
- model-external dispatch and compatibility `.generate` dispatch agree.

## Required adversarial tests

Kill at least:

- adapter invokes MDLM after selecting block decode;
- requested steps recorded while a different count executes;
- missing required attention mask becomes silent `None`;
- a renamed private parameter breaks generation only at runtime;
- explicit unsupported selection silently falls back to `auto`;
- hybrid model is allowed through block decode;
- plugin algorithm registered in one hub resolves through the default hub;
- DiffusionGemma adapter re-enters Unturtle dispatch recursively;
- upstream native config is converted to a masked config;
- class swap or instance-method deletion is still required after the migration;
- output parity passes while RNG/NFE/metadata differs;
- external dispatch and model method choose different algorithms.

## Relationship to #157

#157 may continue reference audit, baseline measurement and candidate selection on its frozen implementation.

Rules during this refactor:

- do not change #157 timing/quality semantics inside infrastructure PRs;
- do not claim removal of dispatch overhead/class swap as a cache/topology candidate win;
- if #157 selects a candidate that needs a new runner, register it through this explicit adapter boundary;
- the runner refactor must preserve the existing baseline implementation until a paired benchmark intentionally replaces it.

## Public compatibility and deprecation

During this issue:

- family-specific generation configs stay public at their current paths;
- current model `.generate` behavior remains covered;
- `algorithm=` compatibility is not removed without a separate deprecation notice;
- new code should prefer the explicit external dispatch boundary after it is stable;
- no new generation mixin is required solely to participate in registry dispatch.

## Stop conditions

Stop/re-scope if:

- a common result requires flattening method-specific semantics;
- a candidate adapter has only one pack-local consumer and needs core fields for convenience;
- output parity cannot be established because the old path itself is nondeterministic—freeze the exact diagnostic instead of guessing;
- DiffusionGemma cannot be invoked without class mutation due to a confirmed upstream bug; isolate that incompatibility and decide whether to maintain a narrow compatibility wrapper rather than restoring a generic class-swap system;
- the change begins rewriting sampling algorithms instead of their invocation boundary.

## Acceptance criteria

- [ ] #184 generation execution map is frozen first.
- [ ] Every builtin algorithm has an explicit adapter/runner.
- [ ] Signature guessing is removed from builtin execution.
- [ ] `SupportResult` distinguishes supported/unsupported/unverified with attributable reasons.
- [ ] `GenerationResult` records selected runner and actual execution metadata without universal method semantics.
- [ ] Explicit model-external dispatch works with supplied/default isolated hubs.
- [ ] Existing model `.generate` paths remain behaviorally compatible or have an explicit staged deprecation.
- [ ] DiffusionGemma runs through upstream/native delegation without runtime `__class__` replacement.
- [ ] No instance-level `generate` deletion is required for DiffusionGemma.
- [ ] Plugin runners use the same hub/validation path as builtins.
- [ ] Output/RNG/NFE/config parity is proven per migrated family.
- [ ] #157 benchmark semantics remain unchanged.
- [ ] No universal generation config/state/solver abstraction added.
- [ ] Focused mutation tests + full applicable suite green.

## Implementer order

1. Consume #184's algorithm -> method execution map.
2. Add typed support/result envelopes without moving behavior.
3. Migrate masked algorithms first, one adapter at a time.
4. Migrate research/continuous builtins without promoting capabilities.
5. Add external dispatch and make compatibility paths share selection.
6. Replace DiffusionGemma's class swap only after direct upstream parity is recorded.
7. Delete signature guessing last.

**Comment by nishide-dev** [comment]
## Replacement ready: #203 (closes this issue on merge)

Head `986afae`, artifact digest `b944bbb8…`.

**Signature guessing → explicit runner**: `_call_sampling_loop` invokes every masked loop through the declared contract `loop(input_ids, attention_mask=..., generation_config=..., **options)` — no `inspect.signature`, no positional None-filling; a signature-hiding wrapper can no longer change the binding (the #184-frozen failure, now a discriminating test). Dream's `_sample` hook params defaulted (the one non-uniform loop). **Seven route differentials vs the verbatim main guessing invoker** (Dream/TinyA2D/LLaDA × mdlm/block_decode/bd3lm; cache, config, kwargs, attention mask) — outputs **bit-identical**.

**Runtime `__class__` swap removed; wrapper-first load owns the class**: `_apply_post_load_class_swap`'s three repair jobs (class stamp, unsloth instance-generate pop, generation_config restore) existed because FastModel was the wrong route for wrapper families — those model types now skip FastModel and load through the wrapper-first Auto chain (`generation_config.json` arrives via the ordinary postamble; the removed restore logic is provably dead). No module assigns `__class__` anywhere (AST-asserted). The DiffusionGemma shim routes through `dispatch_generation` (a replaced runner IS what executes); `_generate_canvas` is the single verbatim upstream call site; a plain upstream instance runs its class-level generate — instance shims can never hijack it, and the façade routes un-wrapped canvas instances through the runner instead of TypeError-ing (review fix).

**Mandated mutants all killed**: 旧guessing復活（verbatim main実装を復元→signature-hiding testが誤bind検出）・旧swap復活（`__class__ =` をloaderへ再挿入→AST invariantが失敗）・runner bypass（shimがsuper().generate直呼び→dispatch sentinel未到達）— plus canvas-instance-generate, canvas-target-ignored, wrapper-first-routing-dropped, attention-mask-dropped, and the two review-fix mutants: **9/9**.

**Non-regressions pinned**: class identity（wrapper-first loadがwrapperを産む・restamp不在）・save/reload（persistence 73 on the regenerated artifact）・AutoModel registration（wrapper-first ordering test、loaderへ再ポイント）・adapter recursion（loader suite; 再帰base loadも同じrouting）。4-bit is retained on the wrapper-first route（bnb quantization_configがAuto chainへ流れる — pinned by test; unsloth backendからtransformers/bnb backendへの意図的変更として記録）。

Artifact: the two `REPLACE → #186` verdicts → **REPLACED**; the three swap-owned ledger rows removed with the mutations they described; scanner 0 unclaimed; `--check` deterministic. Gates: broad subset 1885 · architecture+persistence 73 · full fast suite 2773 (only the #192 pair) · CI Lint pass.

(Process note, recorded for honesty: the first battery round produced two false SURVIVEDs because an earlier test-file edit had accidentally deleted the killing tests together with an obsolete class — restored, battery re-run clean.)
